Cleanup of Materials output system

shikhar413 opened the pull request:
<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

## Reason
<!--Why do you need this feature or what is the enhancement?-->

## Design
<!--A concise description (design) of the enhancement.-->

## Impact
<!--Will the enhancement change existing APIs or add something new?-->


<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

refs #29816, #29787, #29781, #29780, and #29819

shikhar413 committed:
General cleanup of Material Output system

- Fix faulty logic when materials are outputted to different files
- Make sure setting outputs = 'all none' results in an error
- Confirm that outputs = "exodus", outputs = "all", and outputs = "none" work as expected

shikhar413 committed:
Add unit tests for checking expected behavior in presence of multiple output files and expected error messages with outputs parameter

shikhar413 committed:
Update documentation on Materials output system, refs #29816, #29787, #29781, #29780, #29819

GiudGiud commented on framework/doc/content/syntax/Materials/index.md:
```diff
@@ -207,7 +207,14 @@ vector or tensor value.
+creates two additional variables called "mat1" and "mat2" that will show up in the output file. In this
+example, the `exodus` name is a special keyword used to signal to MOOSE that the material properties
+should be outputted to the output object created when setting `Outputs/exodus=true`. If multiple output
+Exodus objects exist in the `[Outputs]` block, one or more names can be provided to the "outputs"
```
```suggestion
[Exodus.md] objects exist in the `[Outputs]` block, one or more names can be provided to the "outputs"
```

GiudGiud commented on framework/doc/content/syntax/Materials/index.md:
```diff
@@ -207,7 +207,14 @@ vector or tensor value.
+should be outputted to the output object created when setting `Outputs/exodus=true`. If multiple output
+Exodus objects exist in the `[Outputs]` block, one or more names can be provided to the "outputs"
+parameter. In addition, the reserved output name `all` can be used to output the material property to
+all Exodus objects in the `[Outputs]` block, while the reserved output name `none` can be used to
```
split the text between the example showing an exodus case and 'all'/'none' applying beyond just exodus

shikhar413 replied:
Done

GiudGiud commented on framework/include/outputs/OutputWarehouse.h:
```diff
@@ -61,8 +61,10 @@ class OutputWarehouse : protected PerfGraphInterface
   /**
    * Returns true if the output object exists
    * @param name The name of the output object for which to test for existence within the warehouse
+   * @param is_exodus Optional parameter to check if the output object associated with name must be
+   *                  Exodus type
```
why are we targetting exodus? Let's find a templated way

GiudGiud commented on framework/src/actions/CheckOutputAction.C:
```diff
@@ -103,8 +103,8 @@ CheckOutputAction::checkMaterialOutput()
-    // Check that the outputs exist
-    _app.getOutputWarehouse().checkOutputs(outputs);
+    // Check that the outputs exist, and that they are of Exodus type
+    _app.getOutputWarehouse().checkOutputs(outputs, /* exodus = */ true);
```
but we could want to be outputting to nemesis / VTK outputs

GiudGiud commented on framework/src/outputs/OutputWarehouse.C:
```diff
@@ -120,9 +120,18 @@ OutputWarehouse::addOutput(std::shared_ptr<Output> const output)
 
 bool
-OutputWarehouse::hasOutput(const std::string & name) const
+OutputWarehouse::hasOutput(const std::string & name, const bool is_exodus) const
```
I d rather you  add a templated version over this boolean

GiudGiud commented on test/tests/materials/output/gold/output_multiple_files_exodus1.e:
are the regolds unavoidable? can you add a thorough justification in the commit description

shikhar413 replied:
I wanted to rename these gold files since they are used for other tests that I added and wanted them to have a more descriptive name

shikhar413 replied:
This is being captured as a file removal and new file addition in git instead of a rename, let me see why this is the case

GiudGiud replied:
> This is being captured as a file removal and new file addition in git instead of a rename, let me see why this is the case

ah yeah that can happen. it s just the platform showing that

GiudGiud commented on test/tests/materials/output/tests:
```diff
@@ -68,9 +68,53 @@
+                  Outputs/exodus2/file_base='output_exodus2_mat_block2'"
+      detail = 'outputting certain properties within the material definition to different files;'
+    []
+    [multiple_materials_single_file]
```
these tests are great

GiudGiud reviewed (changes requested):
I dont know if the MaterialPropertyOutput works as expected with VTK/Nemsis, but even if it does not we dont want to further make it only work with exodus

shikhar413 commented on framework/doc/content/syntax/Materials/index.md:
```diff
@@ -207,7 +207,14 @@ vector or tensor value.
 Output of `Material` properties is enabled by setting the "outputs" parameter. The following example
-creates two additional variables called "mat1" and "mat2" that will show up in the output file.
+creates two additional variables called "mat1" and "mat2" that will show up in the output file. In this
+example, the `exodus` name is a special keyword used to signal to MOOSE that the material properties
```
That’s a good point, I’m not sure let me check

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2675986/), step Generate coverage on b9b857f wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>e5db19</th>
      <th colspan=4>#29820 b9b857</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.24%</td>
      <td>85.24%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>108491</td>
      <td>108515</td>
      <td>+24</td>
      <td>34</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18793</td>
      <td>18794</td>
      <td>+1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/29820/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29820/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/29820/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/29820/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/29820/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/29820/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/29820/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/29820/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/29820/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/29820/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/29820/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/29820/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/29820/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/29820/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/29820/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/29820/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/29820/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/29820/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/29820/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/29820/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/29820/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/29820/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/29820/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/29820/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/29820/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/29820/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/29820/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/29820/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/29820/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/29820/coverage/xfem)

</details>



This comment will be updated on new commits.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2675939/), step Docs: sync website on b9b857f wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/29820/site)

This comment will be updated on new commits.

GiudGiud commented:
Does the selection of which properties to output rely on the hide/show parameters of AdvancedOutput by any chance? 
This output base class is widely used and could be a filter on which output to support

>  if this is possible without hardcoding the datatypes where this is supported,

I expect we'd need to hardcode something like "outputs = exodus" but something like outputs = 'some_name' or 'all' shouldnt need to?

> then the material output functionality is supported for that data format

there should be some mechanism to further restrict these auxvariables, so they dont appear in every output when specified to only be output in one

shikhar413 commented:
> Does the selection of which properties to output rely on the hide/show parameters (and subsequent attributes) of AdvancedOutput by any chance?

I'm not seeing any special logic in MaterialOutputAction that is linked to AdvancedOutput. The way I understand the code, the aux variable is created regardless of the type of output object linked to the material, and then at the point of outputting aux variables to the output file, the function `addInterfaceHideVariables` is called to prevent the variable from being outputted if the material isn't linked to that specific output name

shikhar413 commented:
Something I'm noticing in the public app tests is that a lot of Malamute tests are failing because material outputs are being used on a csv output object. We should confirm that Materials output indeed doesn't do anything for csv output object and remove `Materials/*/output = csv` from any input files that try to do this.

shikhar413 commented:
> there should be some mechanism to further restrict these auxvariables, so they dont appear in every output when specified to only be output in one

I'm looking at the output types that inherit AdvancedOutput and it looks like most output types inherit this class, so trying to restrict the name of outputs to be of type AdvancedOutput won't be restricting the allowable output formats all that much

shikhar413 commented:
Following up here - any idea how we'd want to restrict materials output based on the output object type? It doesn't seem like there's a general way to tell if a specific output type supports material data output or not. The missing piece here is:

Does the output type support auxkernel output (for example, exodus, vtk). If so, we should enable materials output. Otherwise, we should not support this output data type (for example, console, csv).

This can also be tabled to a future issue. For now, we can allow material data output regardless of output type, as this PR fixes other issues with the materials output system as well

GiudGiud commented:
@shikhar413 

> Does the output type support auxkernel output (for example, exodus, vtk). If so, we should enable materials output. Otherwise, we should not support this output data type (for example, console, csv).

I think we should just add a routine to Output base class that is 
```
bool supportsMaterialPropertyOutput
```
and if it does not, then error on the output.
You could keep CSV as "True" until Malamute is patched.

shikhar413 commented:
> I think we should just add a routine to Output base class that is bool supportsMaterialPropertyOutput() const

I was thinking about this, is there a programmatic way to determine this or would it be hardcoded based on what we know would be non-supported data types (so far csv, console, and gnuplot, not sure about gmv and checkpoint)

GiudGiud commented:
defaults to false.

if someone creates a bnew output class, it's virtual and they can decide true or false

shikhar413 commented:
> defaults to false, hardcoded for the ones we know
if someone creates a new output class, it's virtual and they can decide true or false

Sounds good. I'll work on this

> I could see a very decent CSV output of material properties using a ElementMaterialSampler. But it's not implemented at this time I think ?

It doesn't seem like it, the VPP takes the material name as input and relies on Outputs/csv=true, instead setting Materials/*/outputs=csv

GiudGiud commented:
> It doesn't seem like it, the VPP takes the material name as input and relies on Outputs/csv=true, instead of setting Materials/*/outputs=csv

I dont see anything either. It would be adding a Sampler either on the auxvariable or the material property. 
Looks like both BISON and malamute have a ton of inputs concerned so we cant error on unsupported CSV yet

shikhar413 commented:
@GiudGiud I've submitted a PR to Malamute to remove material output to csv object -  https://github.com/idaholab/malamute/pull/183. Will you be able to review this?

Let me know if you'd be able to get a similar PR into Bison, or if we should set supportsMaterialPropertyOutput to true for csv type

GiudGiud commented:
I dont have merge rights to malamute.
I ll make a PR to bison

shikhar413 committed:
Restrict material output to specific types of output types

GiudGiud commented:
@shikhar413 
one of the BISON failures is guenuine. When setting outputs  = 'exodus' we want every single exodus output to comply and have this output. 
But in the run, only the standard `Outputs/exodus=true` output got the material properties, the secondary output, created with Outputs/type=Exodus did not get it

shikhar413 commented:
@GiudGiud that's by design. when you set `Materials/*/outputs = exodus`, it will output material property to the object created by `Outputs/exodus = true`. If you want to add material output to the secondary output, you will need to set `Materials/*/outputs = 'exodus name_of_second_output'`

shikhar413 commented:
The reason we don't want material output to every exodus output object is because then we don't have a way to restrict material output to just the output object created by `Outputs/exodus=true`

GiudGiud commented:
I can get behind that. I'll see what the BISON folks say

shikhar413 commented:
FYI malamute PR was merged in https://github.com/idaholab/malamute/pull/183

I'll try to get the updates to the material output system in soon

GiudGiud commented:
BISON patch is up 
https://github.inl.gov/ncrc/bison/pull/6317

I think you should only output a warning for now. There was nearly 50 inputs failing and the simulation was doing what they wanted in the end (output a variable, which was not output to CSV but used for other postprocessing that was output to CSV)

shikhar413 commented:
> BISON patch is up
https://github.inl.gov/ncrc/bison/pull/6317

Thanks!

Ok, I can set the value to true for csv but throw a warning in there saying nothing will get outputted to csv by setting `outputs=csv_obj`

GiudGiud reviewed (commented):
You could test all the other formats by adding them to your Outputs block and setting a material to output to them.

we dont check that they are actually in the file, but we check that we do not hit an error from "supportsMaterialPropertyOutput"

shikhar413 commented on test/tests/materials/output/tests:
```diff
@@ -99,15 +144,41 @@
+    expect_err = "When setting output name to reserved name 'none', only one entry is allowed in outputs parameter"
+    cli_args = "Materials/test_material/outputs='none exodus'"
+
+    requirement = "The system shall report an error if multiple output values are provided and one of those values is a reserved name"
```
In this case, reserved names according to MOOSE are all and none. If both are provided or we have something like “all exodus” it’s unclear whether the user wanted all or just exodus, so I have this error out

GiudGiud replied:
so 'exodus' is not a reserved name?

GiudGiud replied:
I would want to allow something like "exodus nemesis"

shikhar413 replied:
That’s allowed since neither of those are reserved names. Those are the names given to the output object when exodus = true or nemesis = true

shikhar413 commented on framework/include/outputs/OutputWarehouse.h:
```diff
@@ -117,10 +119,20 @@ class OutputWarehouse : protected PerfGraphInterface
+   * @return A set of all output names that support material output
    */
-  void checkOutputs(const std::set<OutputName> & names);
+  const std::set<OutputName> getAllMaterialPropertyOutputNames() const;
```
Why not have this cost?

GiudGiud replied:
You should be getting a compiler warning that it s not doing much to return by const value

shikhar413 commented on framework/src/outputs/OutputWarehouse.C:
```diff
@@ -314,11 +323,37 @@ OutputWarehouse::buildInterfaceHideVariables(const std::string & output_name,
+  }
+  if (!reserved_name.empty() && names.size() > 1)
+    mooseError("When setting output name to reserved name '" + reserved_name +
+               "', only one entry is allowed in outputs parameter.");
```
I'm thinking more about this and I think it makes sense to keep this check where it is. Setting outputs = 'exodus all' is always going to be ill-defined, whether this is check on materials output or something else like aux var output

GiudGiud replied:
ok agree

shikhar413 committed:
Restructure code to separate out hasMaterialPropertyOutput() from hasOutput()

shikhar413 committed:
Extend supportsMaterialPropertyOutput to other output object types

shikhar413 commented on framework/include/outputs/OutputWarehouse.h:
```diff
@@ -61,8 +61,10 @@ class OutputWarehouse : protected PerfGraphInterface
+   * with name must support material output
    */
-  bool hasOutput(const std::string & name) const;
+  bool hasOutput(const std::string & name, const bool supports_material_output = false) const;
```
Done

shikhar413 commented on framework/src/outputs/OutputWarehouse.C:
```diff
@@ -314,11 +323,37 @@ OutputWarehouse::buildInterfaceHideVariables(const std::string & output_name,
+  std::set<OutputName> output_names;
+  for (const auto & pair : _object_map)
+  {
+    auto * output = dynamic_cast<Output *>(pair.second);
```
Done

shikhar413 commented:
Addressed all comments, extended `supportsMaterialPropertyOutput` to relevant output object types, and added tests for these different output object types

shikhar413 commented:
@GiudGiud I'm running into a lot of test failures because of the csv warning that was added. Looks like this is a legitimate use case for outputting material data as an aux variable through the `outputs` parameter and using the aux variable to output something else in csv format. Should I remove the warning?

shikhar413 commented:
One option would be to set `allow_warnings = true` to all affected tests, but this would be required at the BISON level as well

GiudGiud commented:
> I'm running into a lot of test failures because of the csv warning that was added. Looks like this is a legitimate use case for outputting material data as an aux variable through the outputs parameter and outputting another postprocessor that depends on the aux variable into the csv file.

it s pretty hacky imo. Make it an info message, we can tolerate the behavior

shikhar413 commented:
> it s pretty hacky imo. Make it an info message, we can tolerate the behavior

I agree, definitely feels like a usage that works but wasn't the originally intended design

shikhar413 committed:
Add additional input parameter to supportsMaterialPropertyOutput to suppress CSV info when called by getAllMaterialPropertyOutputNames()

shikhar413 committed:
Update solid_mechnics tests where csv material output was expected

shikhar413 commented:
Was finding that the mooseInfo statement was getting triggered when Outputs/csv = true even if material output wasn't being outputted to CSV type, because `getAllMaterialPropertyOutputNames` will always get called when Materials/*/outputs is present. Therefore I added a new parameter to supportsMaterialPropertyOutput to suppress the mooseInfo line when it gets called by this routine in this commit - https://github.com/idaholab/moose/pull/29820/commits/d6d41773c3216dfdaad008b643f6e766c5dade7d

Let me know if you have comments on the design. The commeted suppress_warnings variable is added to prevent a bunch of unused parameter warning thrown by the compiler

Also had to update some solid_mechanics tests where it was csvdiff-ing because it expected Materials/*/outputs=csv to be set in the materials block, I'm hoping other codes don't throw a bunch of errors because of the changes in this PR

shikhar413 commented:
Getting a lot of app failures. I need to investigate what is going on here. In theory the aux kernels that are needed for outputting to csv are still defined when Materials/*/outputs=exodus or whatever other output type is chosen for the outputs parameter, so not sure why the code I have right now is requiring that the outputs parameter be set to csv instead.

shikhar413 commented:
Ok, I have a better explanation of what's going on. Previously, when you set Materials/*/outputs = exodus, material output data would still be added to all output objects defined in the [Outputs] block. This was a bug that this PR is trying to fix based on the diffs to MaterialOutputAction.C. However, now that the bug is fixed, the material output is no longer going to the all output objects by default. The tests that are failing are the ones that have something like `Outputs/csv = true` set and had material output being diverted to the csv object unintentionally, and the gold csv file expects the material property data to exist in the csv object. In order for the tests in this PR to pass, therefore, the Materials/*/outputs=csv needs to be set explicitly so that the material output data continues to exist in the csv output like before.

This means that in order for this PR to go through, we'll need to update the failing tests in all inputs where we have this behavior where material output is being directed to a CSV file by mistake. There are two ways to do this:

1. Set Materials/*/outputs=csv in failing tests (see diffs to failing tests in solid_mechanics input files)
2. Set Materials/*/outputs=all in failing tests. As long as the failing test is CSVDiff, any additional output types where material properties are being directed to shouldn't affect the test passing

Thoughts on this @GiudGiud ? I can update the failing tests for all public tests, but I don't have access to BISON and so you'd need to help me with that (PR you opened will probably need to be modified if it hasn't been merged yet)

shikhar413 committed:
Set file_base in output types tests as the name is clashing with other tests for recovery

shikhar413 committed:
Update combined tests to divert CSV material output to the correct object

lynnmunday reviewed (changes requested):
I appreciate this work and I'm glad you are fixing the output bugs.  I think this PR should only cover that case.  

I think a better fix would be to create an option in Materials that would cause them to create the aux variables from a list of properties in that material.  These auxvaribles could then be used by other objects.

I tried to read some of the comments, but what is the plan for csv output and other outputs types that are not supported. Is that just a temporary fix that will later be removed when some other type of dummy output is available.  Allowing outputs=csv seems too hacky since it can't output anything.  Also, I think all of the other outputs types should just be disabled like nemesis etc since they can't be checked.

shikhar413 commented on test/tests/materials/output/tests:
```diff
@@ -92,22 +137,110 @@
 
+  [output_types]
+    requirement = "The system shall support material property output to output objects of type "
+    [csv]
```
I’ve tested them on my own and all those output types output data to the output file but we don’t have a way to test what is written to the output because this can only be done with Exodiff and Csvdiff

shikhar413 replied:
These tests basically demonstrate that these output tests can output material data without failing

shikhar413 replied:
Material csv output doesn’t do anything, it will only output something if there is a post processor defined in the input file that uses the aux variables created by material output csv and the postprocessor can write data to the output csv file

lynnmunday replied:
I don't like the outputs=csv.  I put an example of using using MaterialRealAux in the blackbear PR.  

https://github.com/idaholab/blackbear/pull/409#pullrequestreview-2707595528

It will be annoying to fix them all using the auxvariables but I think that is better than allowing an output type that doesn't do anything.  Once the auxvariables are created from the materials, they will be written to the exodus files and might cause exodiff's.  You can make these new variables hidden in the outputs block to avoid this.

shikhar413 replied:
Let me try this and see if this accomplishes the same result. If so, it will be a matter of updating all tests (mostly BISON) to follow this convention instead of using Materials/*/materials=csv. @GiudGiud thoughts on this? Since I won’t be able to update the bison tests

shikhar413 replied:
I posted on the blackbear PR but the alternative fix recommended by @lynnmunday should work as well, but I'm not sure if this will be able to address all test failures caused by disallowing material output for csv output objects

shikhar413 commented on framework/include/outputs/CSV.h:
```diff
@@ -32,6 +32,8 @@ class CSV : public TableOutput
    */
   CSV(const InputParameters & parameters);
 
+  bool supportsMaterialPropertyOutput(bool suppress_warnings) const override;
```
This variable controls if the mooseInfo line should be suppressed (see CSV.C/h to see what the moose info line is currently, this is basically a statement to the user saying that setting Materials/*/outputs=csv

shikhar413 replied:
This variable has nothing to do with whether variable output is being written or not. Any output type that has supportsMaterialPropertyOuptut to true should in theory be outputting something to the output file

shikhar413 commented:
@lynnmunday I agree with you that the best option would not be to allow csv output. However many apps already make use of the csv material output into a postprocessor that will output to csv, so disallowing this option would cause a lot of app tests to fail (solid_mechanics, combined module, bison, malamute, and blackbear)

Originally I had it set to disallow material output to csv, but this caused too many app tests failures

shikhar413 commented:
@lynnmunday something to also consider is that this PR aims to correct a number of bugs that exist in the materials output system (see referenced issues). The corner case with material csv output being used in CSVDiff tests was something we discovered along the way. The changes proposed here are meant to get the tests to continue to pass without having to regold a number of tests where the material output system is used in a way that it isn’t meant to. My preference here would be to have a PR that retains the existing functionality without a bunch of re-folding, and have the  issue of csv material output be addressed in another PR

shikhar413 committed:
Remove support for csv material output

shikhar413 committed:
Update tests to not rely on materials output system for csv output

shikhar413 commented:
@lynnmunday I've gone ahead and updated the tests in this repo so that the aux kernels have to be defined explicitly four outputting data to csv, instead of relying on the materials output system to do this. A few things that I found from updating these tests that adds tediousness to this process, that @GiudGiud should also be aware of when it comes to updating the BISON tests

- additional parameters may need to be the explicitly defined AuxKernel to get the same output data. The different parameters that might need to be added can be the calls to `outputHelper` in  `MaterialOutputAction::materialOutput`. For example, you will see the test diffs in `solid_mechanics` require that the additional parameter `components` needs to be added for some tests to get the tests to pass.
- the name of the material aux kernel will not always be `MaterialRealVectorValueAux`. The name of the aux kernel will depend on the type of Material that is trying to be outputted. See the different aux kernel types that are supported in `MaterialOutputAction::materialOutput`. For example, the tests in `combined` module required the aux kernel to be of type `MaterialRealAux` instead of `MateiralRealVectorValueAux`

For now, I've set the code to error out if `Materials/*/outputs = csv`, so that we can see which tests need to be updated. However, I will leave it up to @GiudGiud to make the final call on whether this should be an error or an info message, depending on how he wants to update the BISON tests, since I don't have the permissions to see / update these tests

lynnmunday commented on framework/src/actions/CheckOutputAction.C:
```diff
@@ -103,8 +103,8 @@ CheckOutputAction::checkMaterialOutput()
-    // Check that the outputs exist
-    _app.getOutputWarehouse().checkOutputs(outputs);
+    // Check that the outputs exist, and that the output types support material output
+    _app.getOutputWarehouse().checkOutputs(outputs, /* supports_material_output = */ true);
```
See above comment; I don't think this is used.

shikhar413 replied:
Is used in other parts of `CheckOutputAction`

lynnmunday commented on framework/src/outputs/OutputWarehouse.C:
```diff
@@ -314,11 +328,39 @@ OutputWarehouse::buildInterfaceHideVariables(const std::string & output_name,
+    {
+      if (!hasOutput(name))
+        mooseError("The output object '", name, "' is not a defined output object.");
+      if (supports_material_output && !hasMaterialPropertyOutput(name))
```
supports_material_output is always true, I think you can get rid of it.

shikhar413 replied:
Will be false in non-material calls in `CheckOutputAction`

shikhar413 commented on framework/include/outputs/OutputWarehouse.h:
```diff
@@ -117,10 +123,20 @@ class OutputWarehouse : protected PerfGraphInterface
+   * This method will produce an error if any of the supplied names do not exist in
+   * the warehouse. Reserved names are not considered.
+   */
+  void checkOutputs(const std::set<OutputName> & names,
```
Yeah this was relevant when we were outputting mooseInfo statements. Will remove

shikhar413 replied:
Never mind, if you look at CheckOutputAction.C (not just the diffs), there are calls to this function for other structures that support the `outputs` parameter. The optional parameter needs to be set to true only for the case when material output is concerned

shikhar413 commented on test/tests/materials/output/output_boundary.i:
```diff
@@ -6,61 +6,61 @@
 [Variables]
-  [./u]
+  [u]
   [../]
```
Removed

lynnmunday reviewed (approved):
Thanks, you have resolved all of my comments.  I can't mark them resolved.

shikhar413 commented:
FYI @GiudGiud only the BISON tests remain for tests to pass (removed option to output materials to csv)

GiudGiud commented:
passes locally for me. Let s see if rebasing does it

shikhar413 commented:
Amazing, internal apps are passing now. Thank you so much!

GiudGiud committed:
Add a warning when ignoring the material property output parameters
and a test for it
Adapt tests currently doing this mistake

GiudGiud commented on framework/src/actions/MaterialOutputAction.C:
```diff
@@ -176,6 +176,9 @@ MaterialOutputAction::act()
         }
       }
     }
+    else if (output_properties.size() && outputs.empty())
```
nah the problem is outputs is not set and output_material_properties is set

shikhar413 replied:
I'm looking into this a bit more but this condition should be
`else if (output_properties.size())`

when Materials/*/outputs is not specified, the `outputs` variable will have value `none` based on `mat->getOutputs();`, so `output` will never be empty.

shikhar413 replied:
But then if the user specifies `Materials/*/outputs = ''`, then it will enter the `if` condition and not the `else if` condition instead

GiudGiud replied:
lemme amend that

shikhar413 commented on framework/src/actions/MaterialOutputAction.C:
```diff
@@ -176,6 +176,9 @@ MaterialOutputAction::act()
         }
       }
     }
+    else if (output_properties.size() && (outputs.find("none") != outputs.end()))
```
(outputs.find("none") != outputs.end()) condition shouldn't be needed. It will always be the case if it reaches this part of the else if statement

GiudGiud replied:
good call
updated

shikhar413 replied:
Alright, good to go from me

GiudGiud commented:
looks like some public app failures
https://civet.inl.gov/job/2792121/

i ll take Ferret and Racoon

shikhar413 commented:
What is the fix here, is it to add `outputs = all` or remove `output_properties = ...`

GiudGiud commented:
if they test the exodus, remove `output_propertie` because outputs=all will add to the output

GiudGiud commented:
ferret and ba_hun are unrelated. I might be able to fix them anyway

shikhar413 commented:
https://github.com/idaholab/malamute/pull/190

shikhar413 commented:
Malamute good to go

moosebuild commented:
Job [Public apps](https://civet.inl.gov/job/2793508/) on d778e0f : invalidated by @shikhar413

moosebuild commented:
Job [Public apps](https://civet.inl.gov/job/2793508/) on d778e0f : invalidated by @shikhar413

Re-run failing public app tests

shikhar413 commented:
@GiudGiud public app tests passing, tho some raccoon tests might have been overlooked

GiudGiud commented:
just added more inputs in an extra PR

GiudGiud commented:
Raccoon patch is merged and raccoon passes locally (fails on unrelated tests)

GiudGiud commented:
cleanest public app runs in 6 months

openmpi failure unrelated

shikhar413 commented:
Thanks @GiudGiud and also for fixing the public app tests

shikhar413 commented:
In terms of issues that can be closed, the ones listed at the top of this PR as well as #23550 can be marked as completed

GiudGiud commented:
@shikhar413 I suspect this broke next
https://civet.inl.gov/job/2798709/
The valgrind recipe might tell you more

shikhar413 commented:
Looks like some of the material output tests to vtk need to be updated, others seem unrelated. Will wait for all tests to finish before submitting a patch.

GiudGiud commented:
I ll take care of TMAP8 failure here
https://civet.inl.gov/job/2798726/

GiudGiud commented:
I think something is still wrong with Outputs/exodus/output_material_properties
see 
https://github.com/idaholab/TMAP8/pull/251

shikhar413 commented:
Follow-up issues related to the TMAP8 test failures:

#30260 #30261

shikhar413 commented:
I won't be able to get to this immediately but I can work on these when I have some time freed up
